core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -4495,7 +4495,8 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
       sendRequestsLatch.await()
       verify(blockStoreClient, times(2))
         .finalizeShuffleMerge(any(), any(), any(), any(), any())
-      assert(sentHosts === Seq("hostB"))
+      assert(sentHosts.nonEmpty)
+      assert(sentHosts.head === "hostB" && sentHosts.length == 1)
```

Comment:
Ah, okay. Seems like this is flaky because of the timeout:

```
[info] - SPARK-40096: Send finalize events even if shuffle merger blocks indefinitely with registerMergeResults is false *** FAILED *** (106 milliseconds)
[info]   ArrayBuffer("hostB") was empty (DAGSchedulerSuite.scala:4498)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
[info]   at org.apache.spark.scheduler.DAGSchedulerSuite.$anonfun$new$286(DAGSchedulerSuite.scala:4498)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
```
https://github.com/apache/spark/actions/runs/3129263557/jobs/5078150518

I think it was empty when the condition is checked. Later, I think the array is filled at the time when the exception is actually printed out.

Should we increase the timeout, @mridulm and @wankunde?

Reply:
I think this might a consequence of using `MyDAGScheduler` - `scheduleShuffleMergeFinalize` is overridden and runs within the calling thread, instead of async - which this test is expecting.

+CC @otterc, @venkata91

Reply:
Hmmm .. this is actually more flaky than I thought:
- https://github.com/apache/spark/actions/runs/3145115911/jobs/5112006948
- https://github.com/apache/spark/actions/runs/3146198025/jobs/5114387367

Would be great if @wankunde has a chance to take a look ..